File: backend/app/crud/warehouse.py
from datetime import datetime

from sqlalchemy import func, select
from sqlalchemy.orm import Session, selectinload

from app.models.sku import Sku
from app.models.warehouse import Warehouse, Stock, StockLog


# ── 仓库 ──

def create_warehouse(db: Session, tenant_id: int, code: str, name: str, address: str | None = None) -> Warehouse:
    wh = Warehouse(tenant_id=tenant_id, code=code, name=name, address=address)
    db.add(wh)
    db.flush()
    return wh


def list_warehouses(db: Session, tenant_id: int) -> list[Warehouse]:
    return db.scalars(select(Warehouse).where(Warehouse.tenant_id == tenant_id, Warehouse.is_active.is_(True)).order_by(Warehouse.id)).all()


# ── 库存 ──

def get_stock(db: Session, tenant_id: int, warehouse_id: int, sku_id: int) -> Stock | None:
    return db.scalar(
        select(Stock).where(
            Stock.tenant_id == tenant_id,
            Stock.warehouse_id == warehouse_id,
            Stock.sku_id == sku_id,
        )
    )


def get_or_create_stock(db: Session, tenant_id: int, warehouse_id: int, sku_id: int) -> Stock:
    s = get_stock(db, tenant_id, warehouse_id, sku_id)
    if not s:
        s = Stock(tenant_id=tenant_id, warehouse_id=warehouse_id, sku_id=sku_id, qty=0)
        db.add(s)
        db.flush()
    return s


def list_stocks(
    db: Session,
    tenant_id: int,
    warehouse_id: int | None = None,
    item_type: str | None = None,
) -> list[Stock]:
    stmt = (
        select(Stock)
        .where(Stock.tenant_id == tenant_id)
        .join(Sku, Sku.id == Stock.sku_id)
        .options(selectinload(Stock.sku), selectinload(Stock.warehouse))
    )
    if warehouse_id is not None:
        stmt = stmt.where(Stock.warehouse_id == warehouse_id)
    if item_type == "material":
        stmt = stmt.where(Sku.code.like("MAT-%"))
    elif item_type == "product":
        stmt = stmt.where(~Sku.code.like("MAT-%"))
    stmt = stmt.order_by(Stock.warehouse_id, Stock.sku_id)
    return db.scalars(stmt).all()


def adjust_stock(
    db: Session,
    tenant_id: int,
    warehouse_id: int,
    sku_id: int,
    change_qty: int,
    biz_type: str,
    biz_id: int | None = None,
    remark: str | None = None,
) -> Stock:
    """调整库存，change_qty 正=入库 负=出库"""
    s = get_or_create_stock(db, tenant_id, warehouse_id, sku_id)
    s.qty += change_qty
    log = StockLog(
        tenant_id=tenant_id,
        warehouse_id=warehouse_id,
        sku_id=sku_id,
        change_qty=change_qty,
        balance_qty=s.qty,
        biz_type=biz_type,
        biz_id=biz_id,
        remark=remark,
    )
    db.add(log)
    db.flush()
    return s


def list_stock_logs(
    db: Session,
    tenant_id: int,
    warehouse_id: int | None = None,
    sku_id: int | None = None,
    item_type: str | None = None,
    offset: int = 0,
    limit: int = 50,
) -> list[StockLog]:
    stmt = select(StockLog).where(StockLog.tenant_id == tenant_id).options(selectinload(StockLog.sku), selectinload(StockLog.warehouse))
    if warehouse_id is not None:
        stmt = stmt.where(StockLog.warehouse_id == warehouse_id)
    if sku_id is not None:
        stmt = stmt.where(StockLog.sku_id == sku_id)
    if item_type == "material":
        stmt = stmt.join(Sku, Sku.id == StockLog.sku_id).where(Sku.code.like("MAT-%"))
    elif item_type == "product":
        stmt = stmt.join(Sku, Sku.id == StockLog.sku_id).where(~Sku.code.like("MAT-%"))
    stmt = stmt.order_by(StockLog.id.desc()).offset(offset).limit(limit)
    return db.scalars(stmt).all()


def sum_stock_qty_by_sku_ids(db: Session, tenant_id: int, sku_ids: list[int]) -> dict[int, int]:
    if not sku_ids:
        return {}
    rows = db.execute(
        select(Stock.sku_id, func.sum(Stock.qty))
        .where(Stock.tenant_id == tenant_id, Stock.sku_id.in_(sku_ids))
        .group_by(Stock.sku_id)
    ).all()
    return {int(sku_id): int(qty or 0) for sku_id, qty in rows}

File: frontend-admin-pro/src/api/warehouse.ts
import { http } from '@/utils/http'
import type { ListResp } from '@/types/api'

export type WarehouseOut = { id: number; code: string; name: string; address: string | null }

export type StockOut = {
  id: number
  warehouse_id: number
  warehouse_name: string | null
  sku_id: number
  sku_code: string | null
  sku_name: string | null
  qty: number
  updated_at: string
}

export type StockLogOut = {
  id: number
  warehouse_id: number
  warehouse_name: string | null
  sku_id: number
  sku_code: string | null
  sku_name: string | null
  change_qty: number
  balance_qty: number
  biz_type: string
  remark: string | null
  created_at: string
}

export const warehouseApi = {
  listWarehouses() {
    return http.request<ListResp<WarehouseOut>>({ url: '/admin/warehouse/warehouses', method: 'GET' })
  },
  listStocks(params: { warehouse_id?: number; item_type?: 'product' | 'material' | 'all' }) {
    const p: any = { warehouse_id: params.warehouse_id }
    if (params.item_type && params.item_type !== 'all') p.item_type = params.item_type
    return http.request<ListResp<StockOut>>({ url: '/admin/warehouse/stocks', method: 'GET', params: p })
  },
  exportStocks(params: { warehouse_id?: number; item_type?: 'product' | 'material' | 'all' }) {
    const p: any = {}
    if (params.warehouse_id) p.warehouse_id = params.warehouse_id
    if (params.item_type && params.item_type !== 'all') p.item_type = params.item_type
    return http.request<Blob>({ url: '/admin/warehouse/stocks/export', method: 'GET', params: p, responseType: 'blob' })
  },
  listLogs(params: { warehouse_id?: number; sku_id?: number; item_type?: 'product' | 'material' | 'all'; offset?: number; limit?: number }) {
    const p: any = { warehouse_id: params.warehouse_id, sku_id: params.sku_id, offset: params.offset, limit: params.limit }
    if (params.item_type && params.item_type !== 'all') p.item_type = params.item_type
    return http.request<ListResp<StockLogOut>>({ url: '/admin/warehouse/logs', method: 'GET', params: p })
  },
}
